### PR TITLE
Handle authorization exceptions when reading files

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://devdiv.visualstudio.com/DevDiv/_queries/edit/1851532

### CODE REVIEW

File: src/Compilers/Core/Portable/FileSystem/FileUtilities.cs [hunk 1]
```diff
@@ -6,6 +6,7 @@
 using System.Collections.Generic;
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/FileSystem/FileUtilities.cs:6]: view with whitspace off.

File: src/Compilers/Core/Portable/FileSystem/FileUtilities.cs [hunk 2]
```diff
@@ -304,17 +305,35 @@ internal static Stream OpenAsyncRead(string fullPath)
             return RethrowExceptionsAsIOException(() => new FileStream(fullPath, FileMode.Open, FileAccess.Read, FileShare.Read, 4096, FileOptions.Asynchronous));
         }
 
-        internal static T RethrowExceptionsAsIOException<T>(Func<T> operation)
+        public static T RethrowExceptionsAsIOException<T>(Func<T> operation)
+            => RethrowExceptionsAsIOException(
+                static operation => operation(),
+                operation);
+
+        public static T RethrowExceptionsAsIOException<T, TArg>(Func<TArg, T> operation, TArg arg)
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/FileSystem/FileUtilities.cs:313]: allows for passing static lambdas in.

File: src/Workspaces/Core/Portable/Workspace/Solution/FileTextLoader.cs [hunk 3]
```diff
@@ -171,29 +172,21 @@ public override async Task<TextAndVersion> LoadTextAndVersionAsync(LoadTextOptio
...
-            {
-                var version = VersionStamp.Create(prevLastWriteTime);
 
-                // we do this so that we asynchronously read from file. and this should allocate less for IDE case. 
-                // but probably not for command line case where it doesn't use more sophisticated services.
-                using var readStream = await SerializableBytes.CreateReadableStreamAsync(stream, cancellationToken: cancellationToken).ConfigureAwait(false);
-                var text = CreateText(readStream, options, cancellationToken);
-                textAndVersion = TextAndVersion.Create(text, version, Path);
-            }
+            var textAndVersion = await FileUtilities.RethrowExceptionsAsIOExceptionAsync(
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/FileTextLoader.cs:176]: primary changes are:

1. RethrowExceptionsAsIOExceptionAsync now surrounds *all* the IO (including the deferred work done in CreateReadableStreamAsync).
2. this was converted to static-lambda form just to be a slightly better citizen about allocatinos.

File: src/Workspaces/Core/Portable/Workspace/Solution/FileTextLoader.cs [hunk 4]
```diff
@@ -171,29 +172,21 @@ public override async Task<TextAndVersion> LoadTextAndVersionAsync(LoadTextOptio
...
+                async static t =>
+                {
+                    using var stream = t.self.OpenFile(bufferSize: 1, useAsync: true);
+                    var version = VersionStamp.Create(t.prevLastWriteTime);
 
-            // Check if the file was definitely modified and closed while we were reading. In this case, we know the read we got was
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/FileTextLoader.cs:185]: extracted to helper, used by both the async and sync paths.

File: src/Workspaces/Core/Portable/Workspace/Solution/FileTextLoader.cs [hunk 5]
```diff
@@ -207,20 +200,25 @@ internal override TextAndVersion LoadTextAndVersionSynchronously(LoadTextOptions
...
-            TextAndVersion textAndVersion;
+            var textAndVersion = FileUtilities.RethrowExceptionsAsIOException(
+                static t =>
+                {
+                    // Open file for reading with FileShare mode read/write/delete so that we do not lock this file.
+                    using var stream = t.self.OpenFile(bufferSize: 4096, useAsync: false);
+                    var version = VersionStamp.Create(t.prevLastWriteTime);
+                    var text = t.self.CreateText(stream, t.options, t.cancellationToken);
+                    return TextAndVersion.Create(text, version, t.self.Path);
+                }, (self: this, prevLastWriteTime, options, cancellationToken));
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Workspace/Solution/FileTextLoader.cs:211]: same as above.  the sync path now wraps opening the file *and* reading it in the exception transformation helper.

File: src/Compilers/Core/Portable/FileSystem/FileUtilities.cs [hunk 6]
```diff
@@ -304,17 +305,35 @@ internal static Stream OpenAsyncRead(string fullPath)
...
             try
             {
-                return operation();
+                return operation(arg);
             }
-            catch (IOException)
+            catch (Exception e) when (e is not IOException)
             {
-                throw;
+                throw new IOException(e.Message, e);
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/FileSystem/FileUtilities.cs:321]: Not certain. Looking at Roslyn it goes way back. Maybe @tmat has context on it?

[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/FileSystem/FileUtilities.cs:321]: Note: all my PR is trying to do is move all exception throwing code under the same wrapper to address the core issue being reported.

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-ide ptal.

**Review by ToddGrun — APPROVED** [review]
:shipit:

**Pull request merged**